### DISCUSSION TITLE
Can the localpdf rule be modfied to get the relative path of the pdf file as the _canonical_uri field value

### DISCUSSION THREAD

**Discussion by arunnbabu81** [buggyj/tiddlyclip-plugin#108] [Q&A]
<img width="576" alt="image" src="https://user-images.githubusercontent.com/67494083/204389943-6c8af2d0-ef93-42c9-a497-99709259bd4a.png">

Can the localpdf rule be modfied to get the relative path of the pdf file as the `_canonical_uri` field value. Does TW support it ?
I am asking this because the local pdf file I embed in tiddlers using absolute path as `_canonical_uri` cannot be accessed in mobile. If relative path is used, may be I can acces such pdf files in mobile also.

<img width="577" alt="image" src="https://user-images.githubusercontent.com/67494083/204390363-9450d63b-d0be-4fd7-9ec8-28bdc853d3dc.png">
i tried this one, but its not working.

**Comment by buggyj** [comment]
The @sublink macro only works if the pdffile is in a subdir relative to the tiddlywiki.

**Comment by arunnbabu81** [comment]
<img width="909" alt="image" src="https://user-images.githubusercontent.com/67494083/204496566-671bf151-a75c-4208-9688-fbcdf11fbf4e.png">
the pdf is stored in the subdirectory of the wikis called files. Wikis are stored in the folder called GSD5.

`file:///Users/apple/OneDrive/My%20wiki/GSD5/files/kwee2007.pdf`  - This is the absolute path.

Below are the rules.

<img width="573" alt="image" src="https://user-images.githubusercontent.com/67494083/204498067-e01f5c9b-fbbe-4b70-9590-6414e5fbb24a.png">

<img width="694" alt="image" src="https://user-images.githubusercontent.com/67494083/204498141-b8363fc7-584b-467a-a8f3-21f68ee917eb.png">

This is how clipping was done.
<img width="883" alt="image" src="https://user-images.githubusercontent.com/67494083/204497900-80c84209-276e-40fc-89d5-6de297c286b9.png">

The tiddler created looks like this.

<img width="708" alt="image" src="https://user-images.githubusercontent.com/67494083/204498655-219c5ad7-59bb-485e-978a-57c00075eb38.png">


I have to click on the link to open the pdf. 
<img width="695" alt="image" src="https://user-images.githubusercontent.com/67494083/204498903-026f21ff-7de3-4ffe-b9b0-515ddc2e6e4b.png">


This is the edit view of the tiddler
<img width="701" alt="image" src="https://user-images.githubusercontent.com/67494083/204499029-2cbc9d9d-4cbc-4e83-bc4e-2a001c2593c7.png">


If the pdf tiddler is reopened, again I have to repeat this. 

If I edit the `_canonical_uri` field like this, then the pdf opens directly without additional clicks.

<img width="710" alt="image" src="https://user-images.githubusercontent.com/67494083/204500487-7e5e3a51-bcc8-4b55-95ad-96ff4c1362a5.png">


Is there any solution?

**Reply by buggyj** [reply]
The orignal localtwpdf rule is what you need to use in this case, the _cononical_uri is converted to a relative url by the `@sublink()` macro
`
"$_canonical_uri":"((*@sublink(@linkURL)*))"`


When you put the cursor over a link its address will be copied to `@linkURL` when you do the clipping.

**Reply by arunnbabu81** [reply]
> "$_canonical_uri":"((*@sublink(@linkURL)*))"

But I am getting the absolute path when I am using that.

**Reply by arunnbabu81** [reply]
I opened the pdf in a separate tab and the used the rule given below to clip and its working

<img width="729" alt="image" src="https://user-images.githubusercontent.com/67494083/204562694-df5adc7e-728a-4c83-8d3c-036cac594a47.png">

**Reply by buggyj** [reply]
couldyou post a screenshot of the result of using the original rule?

**Reply by arunnbabu81** [reply]
> couldyou post a screenshot of the result of using the original rule?

<img width="1475" alt="image" src="https://user-images.githubusercontent.com/67494083/204571084-b8b6d685-9513-4925-8560-e81b02cdfaa5.png">

**Comment by buggyj** [comment]
There is a problem with how the browser is handling spaces in the pathname,

here is a fix:


```

exports.name ="sublink";
exports.run  = function(link,global) {
	var here = (this.document.location.toString().split("#")[0]).replace(/(.*)\/.*?$/,"$1");
	var  pos = -1;
        link = link.replace(/ /g,"%20");
	pos = link.indexOf(here);
	if (!global && pos === 0) return "./" + link.substr(here.length+1);
	return link;
}



```

**Reply by buggyj** [reply]
Although spaces in directory names are permitted I would strongly advise using hyphens instead.

**Reply by arunnbabu81** [reply]
> Although spaces in directory names are permitted I would strongly advise using hyphens instead.

I am ok with pathnames without spaces. Are you referring to the folder name `My Wiki` . I will try after removing the space and report back(when I am back on desktop)

**Reply by arunnbabu81** [reply]
> > Although spaces in directory names are permitted I would strongly advise using hyphens instead.
> 
> I am ok with pathnames without spaces. Are you referring to the folder name `My Wiki` . I will try after removing the space and report back(when I am back on desktop)

It does work after i renamed the folder without spaces. Thank you @buggyj
